# Speech Pre Processing

This notebook is used for:
- Loading audio with librosa
- convert to pytorch tensor
- understand sampling rate, frequency content and how to see the waveform

https://huggingface.co/learn/audio-course/chapter1/preprocessing

In [11]:
import pathlib as Path
# import torch
import librosa as lb
import matplotlib.pyplot as plt

In [15]:
audio = 'Test.flac'
wavefrom, sr = lb.load(audio, sr=None)

# sr is sampling rate. It is used for resampling
# if i do not mention then it will automatically resample to 22050HZ


Now this will make a numpy array with the samples

samples is nothing but values at every value or every timestamp of the waveform

Sampling rate is the sampling rate it is an int value only


In [16]:
print(type(audio))
print(waveform.shape)
print(sr)

<class 'str'>
(114960,)
16000


So this means, 114960 samples are there in the entire thing

16000 means i took 16000 samples per second 


Now let's look into converting into PyTorch Tensor


In [18]:
import torch
torch_waveform = torch.from_numpy(waveform) 
# this means i am getting waveform from numpy array to torch

print(type(torch_waveform))
print(torch_waveform.shape)

<class 'torch.Tensor'>
torch.Size([114960])


Pytorch convention is to take channels and samples

In [ ]:
# to fixc this, we will use waveform.unsqueeze(0)

torch_waveform = torch_waveform.unsqueeze(0)

"""
upon doing this, we will add another dimension to the index
"""

But bro, now if I have an N channel audio file, how do i deal with it? how do i deal in pytorch sense?

Let's go back to librosa, 

In [22]:
audio, sr = lb.load(audio, sr = None, mono=False)
print(audio.shape)

(114960,)


usually this would return like 
(channels, N)

#### let's now understand sampling rate

In [23]:
duration = len(waveform)/sr
print(duration)

7.185


In [26]:
torch_waveform = torch_waveform.unsqueeze(0)
duration = torch_waveform.shape[1]/sr
print(duration)

7.185


Let's now design a Robust loader for any sample size, for any channels 

In [27]:
def load_audio(path):
    audio,sr = lb.load(path, sr=None, mono=False)

    if audio.ndim==1:
        waveform = torch.from_numpy(audio).unsqueeze(0) # then my shape will come (1, Samples)
    
    else:
        waveform = torch.from_numpy(audio)
    
    return waveform, sr